extensions/migrate.md

```diff
@@ -0,0 +1,321 @@
+# IRCv3 `migrate` Extension
+
+*Copyright &copy; 2016 Alex Iadicicco <https://github.com/aji>*
+
+*Copyright &copy; 2016 Justin Kaufman <https://github.com/akaritakai>*
+
+*Unlimited redistribution and modification of this document is allowed provided
+that the above copyright notice and this permission notice remains intact*
+
+This document describes the `migrate` IRCv3 client capability and associated
+protocol, which allows servers to migrate client sessions between servers, with
+appropriate client cooperation.
+
+Clients that support the migration protocol outlined in this document should
+register the `migrate` client capability. If the server supports client
+migrations (even if the network is not in a configuration to support
+migrations), it should accept the capability.
+
+# Protocol overview
+
+From the client's perspective, a migration works as follows. The details of this
+protocol are described later in this section.
+
+1. The server sends `MIGRATE START` to indicate that it wants to migrate the
+   client to a different server. Included in this message are a *resume token*
+   and *confirm token* that will be used when connecting to the new server.
+
+2. The client confirms the migration request with `MIGRATE OK`. After sending
+   `MIGRATE OK` the client should stop sending normal IRC traffic, as it will
+   not be processed. If for some reason the client does not wish to migrate,
+   then the migration fails. After receiving `MIGRATE OK`, the server can begin
+   coordinating the migration with the new server.
+
+3. After the server has finished coordinating the migration with the new server,
+   it sends `MIGRATE PROCEED` and closes the connection. If the connection is
+   closed before receiving `MIGRATE PROCEED`, the migration fails.
+
+4. On connecting to the new server, instead of using the normal connection setup
+   protocol (`NICK`, `CAP`, etc.) the client sends a `MIGRATE RESUME` message
+   containing this migration's resume token. Refer to later documentation for
+   the exact semantics of this message.
+
+5. The new server sends `MIGRATE CONFIRM`, containing this migration's confirm
+   token. If the confirm token does not match, the migration fails.
+
+6. The client sends `MIGRATE END`, and can start using the connection as it
+   would have before sending `MIGRATE OK`.
+
+Roughly, this looks as follows:
+
+    s1 <- PRIVMSG #chan :message 1
+    s1 -> :s1 MIGRATE START s2 +6697 abc def :migrate now
+    s1 <- PRIVMSG #chan :message 2
+    s1 <- MIGRATE OK abc
+    s1 -> :s1 MIGRATE PROCEED abc
+      (client disconnects from s1, connects to s2:6697 using TLS)
```

Comment:
This doesn't allow client to connect to s2 first, and then disconnect from s1, which will result in some downtime. I don't know how important that is.

Reply:
Hm, I'll have to think about the implications of that one. It should be fine but I want to give it some thought first.